designdocs/batched_messages.md

```diff
@@ -0,0 +1,81 @@
+# Batched Messages
+
+## Table of Contents
+* [Context and Scope](#context-and-scope)
+* [Problem Statement](#problem-statement)
+* [Goals and Non-Goals](#goals-and-non-goals)
+* [Proposal](#proposal)
+* [Benefits and Risks](#benefits-and-risks)
+* [Alternatives and Rationale](#alternatives-and-rationale)
+* [Glossary](#glossary)
+
+## Context and Scope
+This design document describes batched messages and APIs for sending and retrieving them.
+
+## Problem Statement
+In order to reduce costs for announcers of messages on-chain as well as reduce network congestion, announcers collate messages into batches of the same type of message and announce the batch location on-chain, instead of announcing each individual message.
+
+## Goals and Non-Goals
+This specifies how batches are to be announced on chain: what is required and how a batch may be partially verified based on on-chain information.
+
+This document does not describe the messages that are supported for batch announcements. Please see the [DSNP Specification](https://spec.dsnp.org/).
+
+This document does not discuss validation of batch file format located at the URL in the announcement, since
+the file format cannot be verified on-chain. For details about batch file format, see [DSNP Spec: Batch Publications](https://spec.dsnp.org/DSNP/BatchPublications).
+
+## Proposal
+* All names are placeholders and may be changed.
+* Errors in the extrinsic(s) must have different, reasonably-named error enums for each type of error for ease of debugging.
+
+### Constants
+* `BatchPageSizeMax` - the maximum number of batch announcements that will be returned by the `get_batches` query
+* `BatchSizeMinBytes` - the minimum possible size of a valid batch file with 1 row
+
+### Enums
+* `MessageType` - all values specified in [DSNP Spec Announcement Types](https://spec.dsnp.org/DSNP/Announcements.html)
+
+### Types
+* `BatchAnnouncementParams<T:Config>`: generic
+    * `batch_url`:`Vec<u8>` the URL of the batch file. Must be IPFS and correctly formatted.
+    * `schema_id`: `SchemaId`  the schema id for messages in this batch.  The Schema Id must be for a "batchable" schema.
+    * `file_size`: `usize`, the size of the batch file, used to determine message fee.  Must be &gt;= the minimum possible DSNP batch file size.
+    * `file_hash`: `<T::Hash>`, the hash of the batch file. Must not be 0 hash
+    * `msa_id`: `MsaId`, the id to use for the announcer.  Must exist and be active
```

Comment:
I don't think this is the id of the announcer, it is the id of the claimed origin (although for batches usually I would expect that to be the announcer, but I could imagine a schema that requires a single user to announce a batch of stuff). DSNP uses batches that are multi-user (and derived by the DSNP Id). Other schemas might not.

Reply:
`batch_url` as a Vec<u8> makes sense, but "Must be IPFS" seems under-specified.

Do we want to accept `https://ipfs.io/ipfs/<cid>`?  If so, do we really mean only get that from this specific gateway, which would be... weird.

Or do we mean `ipfs://<cid>` and if so, do we mean to require inclusion of `ipfs://` or do we mean to require *only* the CID?  Or do we mean either is ok?  I prefer just specifying the CID.

Which version of CID are we specifying?  0, 1, both?  IPFS docs currently say " CIDv0 is still used by default for many IPFS operations, so you should generally support v0" but also [switch to v1 for future-proofing](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats).  I prefer supporting only CIDv1.

At any rate, if CIDv2 arrives some years from now, do we want to support it automatically and immediately (which is what the current text implies) or do we want to consider the matter and then amend the spec?  I prefer the latter.

Maybe the text could look more like:  "the IPFS CIDv1 of the batch file".  I'd be happy to justify my preferences if they're not shared by others.

And do we want to support all of CIDv1 or do we want to force something a little more narrow?  I lean toward getting more specific.

Also: I've been slowly coming around to the idea that we should limit URL length to something like 4096. Given that filesystems limit path length, it's hard to imagine a need for longer URLs, even if the URL is <cid>/path.  The big upside here is just having limits on places where people can stuff arbitrary data.

Reply:
If we're accepting _only_ IPFS, there's no reason to take up extra storage with a redundant URL scheme. 

One way to deal with lengths is to pick only one allowed hash algorithm, version, codec, and base encoding and strictly enforce it. Then I think you can either enforce one length or pick a minimum and maximum length.